Fix compiler warnings from utf8.c on 32-bit build

t-a-k opened the pull request:
This set of changes will silence compiler warnings from `Perl_utf8_to_uv_msgs_helper_` function at utf8.c on 32-bit (e.g. x86) build:
```
utf8.c:1819:35: warning: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘int’} and ‘unsigned int’ [-Wsign-compare]
 1819 |             && UNLIKELY(expectlen > OFFUNISKIP(uv)))
      |                                   ^
utf8.c:1895:45: warning: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘ssize_t’ {aka ‘int’} [-Wsign-compare]
 1895 |                 for (unsigned i = curlen; i < expectlen; i++) {
      |                                             ^
utf8.c:2098:34: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘U32’ {aka ‘long unsigned int’} [-Wformat=]
 2098 |                 Perl_croak(aTHX_ "panic: Unexpected case value in "
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 2099 |                                  " utf8n_to_uvchr_msgs() %d", this_problem);
      |                                                               ~~~~~~~~~~~~
      |                                                               |
      |                                                               U32 {aka long unsigned int}
```

---------------------------------------------------------------------------------
* These changes should not change the behavior of perl, so I think that it does not require a perldelta entry.

t-a-k committed:
utf8.c: Change the type of some length variables from SSize_t to Size_t in Perl_utf8_to_uv_msgs_helper_()

Changed the type of `curlen`, `expectlen`, and `avail_len` from `SSize_t`
to `Size_t`, to shut up compiler warnings about comparison between
same-sized signed and unsigned values on 32-bit build.
These variables represent the number of bytes which cannot be negative,
so it should be safe to be unsigned type.

t-a-k committed:
utf8.c: Fixed formatting specifier for croak() in Perl_utf8_to_uv_msgs_helper_

This call of croak() used to attempt to display U32 variable
`this_problem` with %d format specifier, which triggered a compiler
warning on 32-bit build where U32 is typedef'ed to unsigned long.

mauke commented on utf8.c:
```diff
@@ -1634,12 +1634,12 @@ Perl_utf8_to_uv_msgs_helper_(const U8 * const s0,
     const U8 * send = e;
 
-    SSize_t curlen = send - s0;
+    Size_t curlen = send - s0;
```
Why can't this be negative?

tonycoz replied:
It can be negative in non-DEBUGGING builds, where `s0 <= send` is treated as a reported error.  For DEBUGGING builds perl asserts.

t-a-k replied:
Thank you for comments.  I've forgotten a possibility of `s0 > send`.
I pushed a new commit to move this subtraction to the point after checking `s0 < send` to avoid underflow.

jkeenan commented:
Per suggestion from @mauke on #p5p:  We can compare the logs from two different CI runs, one without the patch in this p.r., one with it.

https://github.com/Perl/perl5/actions/runs/12667183737/job/35300550360#step:7:64

vs.

https://github.com/Perl/perl5/actions/runs/12674636619/job/35324408185?pr=22899#step:7:64

This suggests that the patch has the intended effect.  However, I can't evaluate the code *per se*, and @mauke's concern above should be answered.

t-a-k committed:
utf8.c: Postpone pointer subtraction until it turns out to be safe

In Perl_utf8_to_uv_msgs_helper_(),  "curlen = send - s0;" used to be done
earlier in this function, but this subtraction might underflow as
"send >= s0" (that is, "e >= s0") does not necessarily hold true.

Thanks to @mauke and @tonycoz for pointing this out.
